### ISSUE TITLE
Graceful disconnect

### ISSUE THREAD

**Issue by kubasaw** [shamblett/mqtt_client#145]
Hello, 

Apart from the fact that I think that the library is far from human-readable object-oriented code (maybe it is caused by my very small dart experience) it seems to me that there is an error that is not allowing the client to properly disconnect from the broker.

On lines 81-82 of the `lib/src/connectionhandling/mqtt_client_mqtt_connection_handler.dart` file, the current connection status is checked to send the message. 

```dart
    if ((connectionStatus.state == MqttConnectionState.connected) || //LINE 81
        (connectionStatus.state == MqttConnectionState.connecting)) { //LINE 82
... // Message is sent here
    } else {
      MqttLogger.log('MqttConnectionHandler::sendMessage - not connected');
    }
```

and in the `lib/src/connectionhandling/mqtt_client_synchronous_mqtt_connection_handler.dart` file on lines 96-97 the status is changed and followed by an attempt to send a disconnect frame, which for obvious reasons is never sent
```dart
  /// Disconnects
  @override
  MqttConnectionState disconnect() {
    MqttLogger.log('SynchronousMqttConnectionHandler::disconnect');
    // Send a disconnect message to the broker
    connectionStatus.state = MqttConnectionState.disconnecting; //LINE 96
    sendMessage(MqttDisconnectMessage());// LINE 97
    _performConnectionDisconnect(); 
    return connectionStatus.state = MqttConnectionState.disconnected;
  }

  /// Disconnects the underlying connection object.
  void _performConnectionDisconnect() {
    connectionStatus.state = MqttConnectionState.disconnected;
  }
```

And anyway, could you explain to me why the `_performConnectionDisconnect()` from line 98 of `disconnect()` is called, although it does exactly the same as the next (`return`) line ??

**Comment by shamblett** [comment]
Yes this could do with a tidy, I'll do this for the next release or you could submit a pull request to do this if you wish.  I'm not sure why you think the client is 'far from human-readable object-oriented code', you seemed to read it fairly well, maybe you could supply some Dart code you think is OK.

**Comment by kubasaw** [comment]
Hi,  thanks for the reply.
My adventure with Dart has been going on for about two weeks and I don't feel like I was able to write a piece of code without breaking the rest. My thoughts are based on the way of thinking which a have gained when I writing in C for microcontrollers and a bit in Python for GUI apps.

For example, I do not quite understand why a some part of configuration (hostame and clientID) is loaded into the client via constructor, another part by client methods, and others as ConnectionMessage properties or methods.

By the way, when I understand how it all works, I will try to help in the development of the library ;)

**Comment by shamblett** [comment]
Updated, re-published at version 5.6.3

**Comment by kubasaw** [comment]
Thank you very much!
